### DISCUSSION TITLE
[Contract Verification] Verify bytecodehash vs verify contract

### DISCUSSION THREAD

**Discussion by agsola** [zkSync-Community-Hub/zksync-developers#406] [Contract Verification]
When I deploy a contract with deployDeps, the deps code is deployed, but they don't have any contract associated until I use a create or create2.

How can I verify their bytecodehash to allow automatic verification once they are deployed?

**Comment by JackHamer09** [comment]
@agsola Unfortunately at the moment we don't support bytecode based verification. This means each contract should be verified manually after deployment.
You have several options:
- [Verify through Block Explorer UI](https://docs.zksync.io/build/tooling/block-explorer/contract-verification.html#contract-verification)
- [Verify using hardhat](https://docs.zksync.io/build/tutorials/how-to/verify-contracts.html#verifying-contracts-using-hardhat-zksync-verify)

**Reply by agsola** [reply]
Thank you @JackHamer09 .
Unfortunately we use a factory and the users are the ones creating the new contracts. We are a permissionless protocol. 
I guess that there is no easy way to auto-verify the newly created contracts using create2 from the ContractDeployer contract on zkSync, right?

**Reply by JackHamer09** [reply]
Actually I think it should work with [zkSync Etherscan explorer](https://era.zksync.network/)! Should be similar to all other EVM chains. Probably you should verify it once and then all other contracts with same bytecode should be verified automatically. Let us know if it works for you!

**Reply by agsola** [reply]
Thank you @JackHamer09 . Do you have any contact of that Explorer team?
We are really struggling to verify contracts over there. Not only they don't have the latest zksolc compiler version, but even with older versions it seems to be impossible to verify. We've verified on different explorer like regular etherscan, the zkSync explorer, routescan, Blockscout..., but we are unable to do it here.

**Reply by kvrnc** [reply]
Do write into support via https://era.zksync.network/contactus?id=1 and provide the contract address and source code (JSON input used).

**Reply by agsola** [reply]
We did that last week with no response yet.

**Reply by githubdoramon** [reply]
Did you get a response @agsola ?

**Reply by agsola** [reply]
We got a response that they were going to include the new zksolc version @githubdoramon 
However, they didn't replied to the issue that verification was not working on previous versions either.
We tried to reply back, with no response yet...

**Reply by githubdoramon** [reply]
@kvrnc do you mind checking this pls?

**Reply by kvrnc** [reply]
Hey, should be resolved, new zksolc version will be added and you could attempt to verify older contracts. If you have trouble verifying the contracts, do share them to https://era.zksync.network/contactus?id=1 as well.

**Reply by agsola** [reply]
I don't know why you chose this as answer @bxpana , but unfortunately is far from being the solution @kvrnc 

Having the same error ( Error! Unable to generate Contract Bytecode and ABI (General Exception, unable to get compiled [bytecode])).

And new zksolc is still not available.

I've spent so much time with this, re-trying the same thing over and over.

Could I speak with someone from the team?

I've sent the form a few times with no response yet.

**Reply by bxpana** [reply]
@agsola apologies, I thought thins were resolved with what @kvrnc had shared. I'll try to reach out to the team to see what's up if @kvrnc doesn't respond here.

**Reply by kvrnc** [reply]
Was this the contract that you are unable to verify? https://sepolia-era.zksync.network/address/0xAc3c9288B81d55feD14C17AeABe6e68479E08cC6#code

**Reply by agsola** [reply]
yes @kvrnc 
How was it verified?
I always got failures.

**Reply by agsola** [reply]
@kvrnc ??
The compiler version is still not visible. Not 1.4.0 nor 1.4.1

**Reply by kvrnc** [reply]
They are actually supported, but due to a bug they are showing at the bottom. Will patch this. 
![image](https://github.com/zkSync-Community-Hub/zksync-developers/assets/36660375/389a898c-f8d1-45e6-9dcb-d6b9143f6a05)

Would you be able to try sending [verification via API ](https://docs.zksync.network/api-endpoints/contracts#verify-source-code)instead next time if verification fails for you?

**Reply by agsola** [reply]
The information about the verification API is so scarce, not even an example @kvrnc .
Is there at least an example of using it? Can we use the "hardhat-zksync-verify" and point to the verify API?
Do we have to implement our own script to send the API calls?

**Reply by agsola** [reply]
I've been able to manually verify it, but the JSON parser should be improved. A few errors I got with the content it didn't expect:

![image](https://github.com/zkSync-Community-Hub/zksync-developers/assets/472079/30130a70-e75d-4f52-b15d-6861049e668f)

![image](https://github.com/zkSync-Community-Hub/zksync-developers/assets/472079/f50b4c1f-6e5f-4ab2-bfbe-b165e21f8373)

**Reply by agsola** [reply]
And finally, the worst part.

After verifying one contract I expected the rest of the contracts with the same bytecode to be automatically verified. Unfortunately, they are not :(
Then, I waited for a new contract to see if only new contracts were verified automatically, but neither are.

So we still have to manually verify every contract even if they share the same bytecode, as we are deploying from a factory with create2 and bytecodehash @kvrnc

**Reply by kvrnc** [reply]
Can you share the contract you verified that you also expect the other contracts to similarly verify?

**Reply by agsola** [reply]
Sure @kvrnc 
Example of verified contract: 0xf18fff30792a5ee2d09b6856e65dc8d3cf813655
Same new contracts created after verifying that contract that share the same bytecode:
0x191f0ca2cd74868bc5ce08071ef0269892f79f40
0x96c098130f3c8cc022241205153656a23dd0fd36
0x235e12ef8494abf439c8cf345a4b3f6b87c14ca1

**Comment by agsola** [comment]
@bxpana , has there been any change? Why do you keep closing the ticket?

**Reply by kvrnc** [reply]
Hi @agsola , the contracts are similarly matched. Any further issues you facing?

**Reply by agsola** [reply]
@bxpana closed the ticket with no further information.
Last time they were not. I've just seen they are now matched. Thank you.

**Reply by kvrnc** [reply]
All is good 🙏🏻

**Reply by bxpana** [reply]
@agsola apologies for that, I thought it was resolved but glad we got everything figured out now.
